server/logging/firehose.go

```diff
@@ -0,0 +1,162 @@
+package logging
+
+import (
+	"encoding/json"
+	"math"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/firehose"
+	"github.com/aws/aws-sdk-go/service/firehose/firehoseiface"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+)
+
+const (
+	firehoseMaxRetries = 8
+
+	// See
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/firehose/#Firehose.PutRecordBatch
+	// for documentation on limits.
+	firehoseMaxRecordsInBatch = 500
+	firehoseMaxSizeOfRecord   = 1000 * 1000     // 1,000 KB
+	firehoseMaxSizeOfBatch    = 4 * 1000 * 1000 // 4 MB
+)
+
+type firehoseLogWriter struct {
+	client firehoseiface.FirehoseAPI
+	stream string
+	logger log.Logger
+}
+
+func NewFirehoseLogWriter(region, id, secret, stream string, logger log.Logger) (*firehoseLogWriter, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(id, secret, ""),
+		Region:      &region,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "create Firehose client")
+	}
+	client := firehose.New(sess)
+
+	f := &firehoseLogWriter{
+		client: client,
+		stream: stream,
+		logger: logger,
+	}
+	if err := f.validateStream(); err != nil {
+		return nil, errors.Wrap(err, "create Firehose writer")
+	}
+	return f, nil
+}
+
+func (f *firehoseLogWriter) validateStream() error {
+	out, err := f.client.DescribeDeliveryStream(
+		&firehose.DescribeDeliveryStreamInput{
+			DeliveryStreamName: &f.stream,
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "describe stream %s", f.stream)
+	}
+
+	if (*(*out.DeliveryStreamDescription).DeliveryStreamStatus) != firehose.DeliveryStreamStatusActive {
+		return errors.Errorf("delivery stream %s not active", f.stream)
+	}
+
+	return nil
+}
+
+func (f *firehoseLogWriter) Write(logs []json.RawMessage) error {
+	var records []*firehose.Record
+	totalBytes := 0
+	for _, log := range logs {
+		// We don't really have a good option for what to do with logs
+		// that are too big for Firehose. This behavior is consistent
+		// with osquery's behavior in the Firehose logger plugin, and
+		// the beginning bytes of the log should help the Fleet admin
+		// diagnose the query generating huge results.
+		if len(log) > firehoseMaxSizeOfRecord {
+			level.Info(f.logger).Log(
+				"msg", "dropping log over 1MB Firehose limit",
+				"size", len(log),
+				"log", string(log[:100])+"...",
+			)
+			continue
+		}
+
+		// If adding this log will exceed the limit on number of
+		// records in the batch, or the limit on total size of the
+		// records in the batch, we need to push this batch before
+		// adding any more.
+		if len(records) >= firehoseMaxRecordsInBatch ||
```

Comment:
Should these be `>` instead of `>=`?

Reply:
Line 96 needs to be `>=` otherwise we will end up trying to write one more than the max number of records in a batch. Line 97 could be changed to `>` to ensure we write up to the absolute maximum in a single batch.